### DISCUSSION TITLE
Local Reconstructions Proposal

### DISCUSSION THREAD

**Discussion by v-kylegre** [gadgetron/gadgetron#1162] [Ideas]
# Local Reconstructions Proposal

There's been some desire to extend gadgetron's ability to perform reconstructions with local data, meaning that no network socket is required. This is intended to be a draft design and point of discussion for such an implementation.

## High Level Overview

We propose adding an unnamed pipe interface to Gadgetron (optionally input/output files) that would look roughly as follows:

    hdf5_to_stream -f testdata.h5 | gadgetron -c default.xml <additional args?>| stream_to_hdf5 -o file_out.h5

With files:

    hdf5_to_stream -f testdata.h5 -o stream.dat
    gadgetron_cli -c default.xml -i stream.dat -o img_out.dat
    stream_to_hdf5 -i img_out.dat -o file_out.h5

Supporting both methods of invoking Gadgetron is important because [apparently unnamed pipe performance on macOS is less than ideal](https://lemire.me/blog/2021/08/03/how-fast-can-you-pipe-a-large-file-to-a-c-program/). For those who desire the best I/O performance with local files they can consider using a ram disk or other fast storage solution - though we do not forsee I/O being a bottleneck as gadgetron is memory/compute limited (assuming no spinning platters).

## Implementation Details

To implement this we think creating an intermediary binary format leveraging "adaptors" to convert hdf5->binary and vice versa would be the most flexible. Initial thoughts are that the file format would mirror the existing messaging format that goes over the wire. In this case it would be stored in a flat file and pipeable utilising unnamed pipes. Gadgetron would read this flat file much like it reads from a network socket. It's just a stream like any other.

    +------+    +---------------+    +-----------+    +----------------+    +------+
    | hdf5 |--->| input_adapter |--->| gadgetron |--->| output_adapter |--->| hdf5 |
    +------+    +---------------+    +-----------+    +----------------+    +------+

One of the reasonings for this approach is that reading HDF5 directly within gadgetron would couple Gadgetron to the format and that would be a maintainability concern. Now there is already a form of loose coupling within Gadgetron (namely the gadgets which produce ISMRMRD out) but ideally we'd like to avoid the dependency. We think by doing so we can achieve more flexibility by introducing some new tools ("adaptors") to perform a conversion from HDF5 to a binary format and vice versa. We see the adaptors existing within Gadgetron for now, though they could be held in their own repo or perhaps as part of ISMRMRD. The binary format would look simple as follows (the same as what goes over the network):

    +----------------------------+
    |    ISMRMRD HDR             |
    +----------------------------+
    | ACQUISITIONS/WAVEFORM MSGS | (and/or other input messages)
    +----------------------------+

Now I touched on greater flexibility - and that comes by nature of the input/output format being an interface. One could write a new adaptor to say consume data coming that is in an arbitrary and/or proprietary format and as long as that data is convertible to the binary format then that data can be consumed by gadgetron. Same on the output end, the output data can be translated to any desirable format as long as one creates an adaptor to do so. Use cases perhaps being proprietary data format from different scanners, or data originating from different types of sources (storage account perhaps?). These adaptors encapsulate the translation to Gadgetrons binary format and they could be implemented whatever which way is required.

From a code point of view the socket connection is merely used to configure the loader to obtain a processable stream:

    +--------+    +------------+    +------------------+    +------------------+    +------------------+    +--------+
    | Server |--->| Connection |--->| ConfigConnection |--->| HeaderConnection |--->| StreamConnection |--->| Loader |
    +--------+    +------------+    +------------------+    +------------------+    +------------------+    +--------+

This means the entire socket chain can be bypassed with a select few command line arguments added to Gadgetron, and then constructing the loader directly - something like this:

    ...
    Gadgetron::Core::StreamContext ctx(<ismrmrd_hdr>, <paths>, <args>, <storage_address>, <storage_spaces>);
    Gadgetron::Server::Connection::Loader loader(&ctx);

    std::ifstream file(<config_xml_path>, std::ios::in | std::ios::binary);
    config = Gadgetron::Server::Connection::parse_config(file);

    stream = loader.load(config.stream);

    Gadgetron::Core::ChannelPair input_channel;
    Gadgetron::Core::ChannelPair output_channel;
    stream->process(std::move(input_channel.input),
                    std::move(output_channel_.output),
                    ErrorHandler());
    ...

Implementing a command line switch or detecting an open pipe could be the method of switching between the server implementation and the direct command line version.

**Comment by hansenms** [comment]
This looks good IMO. For future proposal, we could try to use the [discussions](https://github.com/gadgetron/gadgetron/discussions) for this kind of proposal and then convert to issue. No need to now. 

Couple of issues to discuss:
* Logging in the Gadgetron is currently done to stdout. What would we do about that?
* In the code above, it is a bit unclear when you read from the input stream/file. You need the ISMRMRD header in order to set up the context, so you would have to start consuming from that first?
* Is this a new binary or the same as the current `gadgetron` binary. Would be great if it could be the latter, but not sure if that ends up being messy.
